File: src/go_tutorials_core_async/tut106.go
package main

import (
  "fmt"
  "time"
)

func f(left, right chan int) {
  left <- 1 + <-right
}

func main() {
  const n = 100000
  leftmost := make(chan int)
  right := leftmost
  left := leftmost
  for i := 0; i < n; i++ {
      right = make(chan int)
      go f(left, right)
      left = right
  }

  start := time.Now()
  go func(c chan int) { c <- 1 }(right)
  fmt.Println(<-leftmost)
  fmt.Println(time.Now().Sub(start))
}

File: src/go_tutorials_core_async/tut106.clj
;; http://talks.golang.org/2012/concurrency.slide#39

(ns go-tutorials-core-async.tut106
  (:use [clojure.core.async]))

(defn f [left right]
  (go (>! left (inc (<! right)))))

(let [leftmost (chan)
      rightmost (loop [n 100000 left leftmost]
                  (if-not (pos? n)
                    left
                    (let [right (chan)]
                      (f left right)
                      (recur (dec n) right))))]
  (time
   (do
     (>!! rightmost 1)
     (println (<!! leftmost)))))

;; "Elapsed time: 1184.47656 msecs"

;; $ go run tut106.go
;; 67.87ms

;; that makes core.async an order of magnitude slower
